File: Prog. Concu y Paralela/Tarea 4/Dekker/TiendaDekker.py
from threading import Thread
from typing import List
import random
import time

# variables compartidas para Dekker
flag: List[bool] = [False, False]
turno: int = 0

class Dekker:
    @staticmethod
    def entrar_seccion(hilo_id):
        global flag, turno
        flag[hilo_id] = True
        otro = 1 - hilo_id

        while flag[otro]:
            if turno != hilo_id:
                flag[hilo_id] = False
                while turno != hilo_id:
                    time.sleep(0) #yield en python
                flag[hilo_id] = True

    @staticmethod
    def salir_seccion(hilo_id):
        global flag, turno
        turno = 1 - hilo_id
        flag[hilo_id] = False

class Tienda:
    
    def __init__(self):
        self.stock = {
            "Camisa Azul chica": 2,
            "Pantalon verde mediano": 3,
            "Tennis negros chicos": 5,
        }
        self.prendas = list(self.stock.keys())

    def comprar(self, num_prenda):
        prenda = self.prendas[num_prenda]
        cantidad = self.stock[prenda]

        if cantidad == 0:
            return [prenda, None]

        self.stock[prenda] = cantidad - 1
        return [prenda, str(cantidad - 1)]

    def get_size(self):
        return len(self.stock)

    def __str__(self):
        output = ["Prenda\tCantidad disponible"]
        for prenda, cantidad in self.stock.items():
            output.append(f"{prenda}:\t{cantidad}")
        return "\n".join(output)

class Cliente(Thread):
    def __init__(self, client, hilo_id, tienda, dekker):
        super().__init__() #clase Thread de python
        self.dekker = dekker
        self.tienda = tienda
        self.client = client
        self.hilo_id = hilo_id

    def run(self):
        random_num = random.randint(0, self.tienda.get_size() - 1)

        self.dekker.entrar_seccion_critica(self.hilo_id)
        try:
            # SECCION CRITICA
            resultado = self.tienda.comprar(random_num)
            # SECCION CRITICA

            if resultado[1] is None:
                print(f"{self.client} quiso comprar" 
                      f"{resultado[0]} pero ya no hay existencias")
            else:
                print(
                    f"- {self.client} ha comprado: {resultado[0]} - "
                    f"Existencias restantes: {resultado[1]}, "
                    f"saliendo de la tienda..."
                )
        finally:
            self.dekker.salir_seccion_critica(self.hilo_id)

def main():
    dekker = Dekker()
    inst = Tienda()
    print(inst, "\n")

    hilos = [
        Cliente("Pedro", 1, inst, dekker),
        Cliente("Luis", 0, inst, dekker)
    ]

    for hilo in hilos:
        hilo.start()

    for hilo in hilos:
        hilo.join()

    print("\n", inst)

main()

File: Prog. Concu y Paralela/Tarea 4/Dekker/TiendaDekker.java
package Dekker;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Random;

public class TiendaDekker {
	// variables compartidas para dekker
	private static volatile boolean[] flag = {
			false, false }; // banderas para indicar que un hilo quiere entrar a la
							// seccion critica
	private static volatile int turno = 0; // var de turno, decide quien de varios hilos entra si ambos lo desean
	
	public final static class Dekker{

		protected void entrarSeccionCritica(int hiloId) {
			flag[hiloId] = true; // el hilo i (actual) tiene la intencion de entrar a
														// la seccion critica

			int otro = 1 - hiloId; // hilo oponente que compite (si el hilo actual es
															// 1, el otro es 0 y viceversa)

			while (flag[otro]) { // si el otro hilo quiere entrar

				// se compite por la seccion critica

				if (turno != hiloId) { // si el turno no le toca al hilo actual entonces espera

					flag[hiloId] = false; // cede el paso hilo oponente

					while (turno != hiloId) // espera su turno
						Thread.yield(); // permite que el otro hilo se ejecute mientras no
														// sea su turno

					flag[hiloId] = true; // vuelve a intentar entrar, se repite el ciclo
				}
			}
		}

		protected void salirSeccionCritica(int hiloId) {
			turno = 1 - hiloId; // cuando el hilo en turno terminal cede el turno a su
													// oponente (1 a 0, 0 a 1)

			flag[hiloId] = false; // da la intencion de que ya no quiere entrar
		}
	}

	public static class Tienda {
		private HashMap<String, Integer> stock = new HashMap<>();
		ArrayList<String> prendas;

		public Tienda() {
			stock.put("Camisa Azul chica", 2);
			stock.put("Pantalon verde mediano", 3);
			stock.put("Tennis negros chicos", 5);
			prendas = new ArrayList<>(stock.keySet());
		}

		public String[] comprar(int numPrenda) {
			String prenda = prendas.get(numPrenda);
			Integer cantidad = stock.get(prenda);

			if (cantidad == 0)
				return new String[] { prenda, null };

			stock.put(prenda,
					cantidad - 1); // quitamos 1 del stock disponible de la prenda
			return new String[] { prenda, String.valueOf(cantidad - 1) };
		}

		public Integer getSize() { return stock.size(); }

		@Override
		public String toString() {
			StringBuilder sb = new StringBuilder();
			sb.append("Prenda\tCantidad disponible\n");
			stock.forEach(
					(k, v) -> sb.append(k).append(":\t").append(v).append("\n"));
			return sb.toString();
		}
	}

	// cliente usando Dekker
	public static class Cliente extends Thread {
		private final Dekker dekker;
		private final Tienda tienda;
		private final String client;
		private final int hiloId;

		public Cliente(String client, int hiloId, Tienda tienda, Dekker dekker) {
			this.hiloId = hiloId;
			this.tienda = tienda;
			this.client = client;
			this.dekker = dekker;
		}

		@Override
		public void run() {
			Integer random = new Random().nextInt(tienda.getSize()); // [0,getSize() )

			dekker.entrarSeccionCritica(hiloId);
			try {

				// SECCION CRITICA
				String[] resultado = tienda.comprar(random);
				// SECCION CRITICA

				if (resultado[1] == null)
					System.err.println(client + " quiso comprar " + resultado[0] +
							" pero ya no hay existencias");
				else
					System.out.println("- " + client + " ha comprado: " + resultado[0] +
							" - Existencias restantes: " + resultado[1] +
							", saliendo de la tienda...");
			} finally {
				dekker.salirSeccionCritica(hiloId);
			}
		}
	}

	public static void main(String[] args) throws InterruptedException {
		Dekker dekker = new Dekker();
		Thread[] hilos = new Thread[2];
		Tienda inst = new Tienda();
		
		System.out.println(inst.toString()); // tienda al principio
		// creamos los hilos alternando entre proceso 0 y 1
		hilos[0] = new Cliente("Pedro", 1, inst, dekker); 
		hilos[1] = new Cliente("Luis", 0, inst, dekker);

		// iniciamos los hilos
		for (Thread hilo : hilos)
			hilo.start();

		// esperamos a que terminen
		for (Thread hilo : hilos)
			hilo.join();

		System.out.println("\n" + inst.toString()); // tienda al final
	}
}